clippy_lints/src/copies.rs

```diff
@@ -225,7 +228,7 @@ fn lint_same_then_else<'tcx>(
             blocks,
             expr,
         );
-    } else if end_eq != 0 || (has_expr && expr_eq) {
+    } else if (end_eq != 0 && !has_expr) || (has_expr && expr_eq) {
```

Comment:
I think the real problem is that `scan_block_for_eq` can return `end_eq > 0 && expr_eq == false`? End statements should not be evaluated if the expressions are not equal.

Reply:
You're right. `scan_block_for_eq` even contained an if block for this purpose, but the condition was not quite correct, it's fixed in the latest version. :upside_down_face: